File: demo/app/src/main/java/com/cyjh/mobileanjian/screencap/ScreenShoterV3.java
package com.cyjh.mobileanjian.screencap;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.hardware.display.VirtualDisplay;
import android.media.Image;
import android.media.ImageReader;
import android.media.projection.MediaProjection;
import android.media.projection.MediaProjectionManager;
import android.os.Handler;
import android.os.HandlerThread;
import android.util.Log;
import android.view.WindowManager;
import java.nio.ByteBuffer;
import p285z2.C4097lg;
import p285z2.C4577yf;

/* JADX INFO: loaded from: classes2.dex */
@TargetApi(21)
public final class ScreenShoterV3 {
    private static final String OooOOo = "ScreenShoterV3";
    private ScreenShotImage OooO;
    private Context OooO00o;
    private MediaProjection OooO0O0;
    private Object OooO0OO;
    private HandlerThread OooO0Oo;
    private Intent OooO0o;
    private Handler OooO0o0;
    private ImageReader OooO0oO;
    private VirtualDisplay OooO0oo;
    private Bitmap OooOO0;
    private int OooOO0O;
    private int OooOO0o;
    private int OooOOO;
    private int OooOOO0;
    private boolean OooOOOO;
    private boolean OooOOOo;
    private volatile boolean OooOOo0;

    public class OooO00o implements ImageReader.OnImageAvailableListener {
        public OooO00o() {
        }

        @Override // android.media.ImageReader.OnImageAvailableListener
        public final void onImageAvailable(ImageReader imageReader) {
            if (ScreenShoterV3.this.OooOOOO) {
                Image imageAcquireLatestImage = ScreenShoterV3.this.OooO0oO.acquireLatestImage();
                if (imageAcquireLatestImage == null) {
                    ScreenShoterV3.this.OooOOo0 = false;
                    return;
                }
                ScreenShoterV3 screenShoterV3 = ScreenShoterV3.this;
                screenShoterV3.OooO = screenShoterV3.OooOO0O(imageAcquireLatestImage);
                ScreenShoterV3.this.OooOO0 = ScreenShoterV3.OooO0O0(imageAcquireLatestImage);
                ScreenShoterV3.this.OooOOo0 = true;
                ScreenShoterV3.OooOOO(ScreenShoterV3.this);
                if (ScreenShoterV3.this.OooO0oO != null) {
                    ScreenShoterV3.this.OooO0oO.setOnImageAvailableListener(null, null);
                }
            }
        }
    }

    public static class OooO0O0 {
        private static final ScreenShoterV3 OooO00o = new ScreenShoterV3(0);

        private OooO0O0() {
        }
    }

    private ScreenShoterV3() {
        this.OooO0OO = new Object();
        this.OooOOOO = false;
        this.OooOOOo = false;
        this.OooOOo0 = false;
    }

    public /* synthetic */ ScreenShoterV3(byte b) {
        this();
    }

    private Bitmap OooO00o(int i, int i2) throws IllegalStateException {
        Image imageAcquireLatestImage = this.OooO0oO.acquireLatestImage();
        if (imageAcquireLatestImage == null) {
            return this.OooOO0;
        }
        Bitmap bitmapOooO0OO = OooO0OO(imageAcquireLatestImage, i, i2);
        this.OooOO0 = bitmapOooO0OO;
        return bitmapOooO0OO;
    }

    public static /* synthetic */ Bitmap OooO0O0(Image image) {
        return OooO0OO(image, 0, 0);
    }

    private static Bitmap OooO0OO(Image image, int i, int i2) {
        Bitmap bitmap;
        try {
            int width = image.getWidth();
            int height = image.getHeight();
            Image.Plane[] planes = image.getPlanes();
            ByteBuffer buffer = planes[0].getBuffer();
            int pixelStride = planes[0].getPixelStride();
            Bitmap bitmapCreateBitmap = Bitmap.createBitmap(((planes[0].getRowStride() - (pixelStride * width)) / pixelStride) + width, height, Bitmap.Config.ARGB_8888);
            bitmapCreateBitmap.copyPixelsFromBuffer(buffer);
            if (width <= height || bitmapCreateBitmap == null) {
                bitmap = bitmapCreateBitmap;
            } else {
                Matrix matrix = new Matrix();
                matrix.setRotate(90.0f, bitmapCreateBitmap.getWidth() / 2, bitmapCreateBitmap.getHeight() / 2);
                float height2 = bitmapCreateBitmap.getHeight();
                float[] fArr = new float[9];
                matrix.getValues(fArr);
                matrix.postTranslate(height2 - fArr[2], 0.0f - fArr[5]);
                Bitmap bitmapCreateBitmap2 = Bitmap.createBitmap(bitmapCreateBitmap.getHeight(), bitmapCreateBitmap.getWidth(), Bitmap.Config.ARGB_8888);
                new Canvas(bitmapCreateBitmap2).drawBitmap(bitmapCreateBitmap, matrix, new Paint());
                bitmapCreateBitmap.recycle();
                bitmap = bitmapCreateBitmap2;
            }
            if (i != 0 || i2 != 0) {
                int width2 = bitmap.getWidth();
                int height3 = bitmap.getHeight();
                Log.e(OooOOo, "load bitmap1 info return  1 width=" + width2 + " height = " + height3);
                Bitmap bitmapCreateBitmap3 = Bitmap.createBitmap(bitmap, 0, i2, width2 + 0, height3 - (i2 * 2), (Matrix) null, false);
                bitmap.recycle();
                bitmap = bitmapCreateBitmap3;
            }
            Log.e(OooOOo, "load bitmap info return  2 width=" + bitmap.getWidth() + " height = " + bitmap.getHeight());
            if (image != null) {
                image.close();
            }
            return bitmap;
        } catch (Exception unused) {
            if (image == null) {
                return null;
            }
            image.close();
            return null;
        } catch (Throwable th) {
            if (image != null) {
                image.close();
            }
            throw th;
        }
    }

    private ScreenShotImage OooO0o0() throws IllegalStateException {
        Image imageAcquireLatestImage = this.OooO0oO.acquireLatestImage();
        if (imageAcquireLatestImage == null) {
            return this.OooO;
        }
        ScreenShotImage screenShotImageOooOO0O = OooOO0O(imageAcquireLatestImage);
        this.OooO = screenShotImageOooOO0O;
        return screenShotImageOooOO0O;
    }

    /* JADX INFO: Access modifiers changed from: private */
    public ScreenShotImage OooOO0O(Image image) {
        try {
            ScreenShotImage screenShotImage = new ScreenShotImage();
            int width = image.getWidth();
            int height = image.getHeight();
            Image.Plane[] planes = image.getPlanes();
            ByteBuffer buffer = planes[0].getBuffer();
            int rowStride = planes[0].getRowStride();
            int pixelStride = planes[0].getPixelStride();
            byte[] bArr = new byte[buffer.capacity()];
            buffer.get(bArr);
            screenShotImage.width = width;
            screenShotImage.height = height;
            screenShotImage.rotation = this.OooOOO;
            screenShotImage.rowStride = rowStride;
            screenShotImage.pixelStride = pixelStride;
            screenShotImage.data = bArr;
            if (image != null) {
                image.close();
            }
            return screenShotImage;
        } catch (Exception unused) {
            if (image == null) {
                return null;
            }
            image.close();
            return null;
        } catch (Throwable th) {
            if (image != null) {
                image.close();
            }
            throw th;
        }
    }

    private synchronized void OooOO0o() {
        recycler();
        updateScreenSize();
        MediaProjectionManager mediaProjectionManager = (MediaProjectionManager) this.OooO00o.getSystemService("media_projection");
        if (this.OooO0O0 == null && mediaProjectionManager != null) {
            this.OooO0O0 = mediaProjectionManager.getMediaProjection(-1, this.OooO0o);
        }
        HandlerThread handlerThread = new HandlerThread("imageReaderHandler");
        this.OooO0Oo = handlerThread;
        handlerThread.start();
        this.OooO0o0 = new Handler(this.OooO0Oo.getLooper());
        String str = C4097lg.OooO00o("getprop phone.id").OooO0O0;
        this.OooO0oO = (str == null || str.length() <= 0) ? ImageReader.newInstance(this.OooOO0O, this.OooOO0o, 1, 3) : ImageReader.newInstance(this.OooOO0O, this.OooOO0o, 2, 3);
        this.OooO0oo = this.OooO0O0.createVirtualDisplay("screen-mirror", this.OooOO0O, this.OooOO0o, this.OooOOO0, 16, this.OooO0oO.getSurface(), null, null);
        this.OooOOOO = true;
        ImageReader imageReader = this.OooO0oO;
        if (imageReader != null) {
            imageReader.setOnImageAvailableListener(new OooO00o(), this.OooO0o0);
        }
    }

    public static /* synthetic */ boolean OooOOO(ScreenShoterV3 screenShoterV3) {
        screenShoterV3.OooOOOO = false;
        return false;
    }

    private boolean OooOOO0() {
        return this.OooOOo0;
    }

    public static synchronized ScreenShoterV3 getInstance() {
        return OooO0O0.OooO00o;
    }

    public final void init(Context context, Intent intent) {
        if (this.OooOOOo) {
            return;
        }
        this.OooOOOo = true;
        this.OooO00o = context.getApplicationContext();
        this.OooO0o = intent;
        this.OooOOO0 = Resources.getSystem().getDisplayMetrics().densityDpi;
    }

    public final void init(Context context, Intent intent, MediaProjection mediaProjection) {
        if (this.OooOOOo) {
            return;
        }
        this.OooOOOo = true;
        this.OooO00o = context.getApplicationContext();
        this.OooO0o = intent;
        this.OooOOO0 = Resources.getSystem().getDisplayMetrics().densityDpi;
        this.OooO0O0 = mediaProjection;
    }

    public final boolean isInited() {
        return this.OooOOOo;
    }

    public final ScreenShotImage obtainScreenShotImage() {
        ScreenShotImage screenShotImageOooO0o0;
        if (this.OooOOo0) {
            try {
                return OooO0o0();
            } catch (IllegalStateException e) {
                e.printStackTrace();
            }
        }
        synchronized (this.OooO0OO) {
            OooOO0o();
            int i = 0;
            while (!this.OooOOo0 && (i = i + 1) < 50) {
                try {
                    Thread.sleep(20L);
                } catch (InterruptedException e2) {
                    e2.printStackTrace();
                }
            }
            screenShotImageOooO0o0 = OooO0o0();
        }
        return screenShotImageOooO0o0;
    }

    public final Bitmap obtainScreenShotImageReturnBitmap(int i, int i2) {
        Bitmap bitmapOooO00o;
        if (this.OooOOo0) {
            try {
                return OooO00o(i, i2);
            } catch (IllegalStateException e) {
                e.printStackTrace();
            }
        }
        synchronized (this.OooO0OO) {
            OooOO0o();
            int i3 = 0;
            while (!this.OooOOo0 && (i3 = i3 + 1) < 50) {
                try {
                    Thread.sleep(20L);
                } catch (InterruptedException e2) {
                    e2.printStackTrace();
                }
            }
            bitmapOooO00o = OooO00o(i, i2);
        }
        return bitmapOooO00o;
    }

    public final void recycler() {
        this.OooOOo0 = false;
        this.OooOOOO = false;
        this.OooOO0o = 0;
        this.OooOO0O = 0;
        this.OooOOO = 0;
        synchronized (this.OooO0OO) {
            HandlerThread handlerThread = this.OooO0Oo;
            if (handlerThread != null) {
                handlerThread.quit();
                this.OooO0Oo = null;
            }
            Handler handler = this.OooO0o0;
            if (handler != null) {
                handler.removeCallbacksAndMessages(null);
                this.OooO0o0 = null;
            }
            MediaProjection mediaProjection = this.OooO0O0;
            if (mediaProjection != null) {
                mediaProjection.stop();
                this.OooO0O0 = null;
            }
            VirtualDisplay virtualDisplay = this.OooO0oo;
            if (virtualDisplay != null) {
                virtualDisplay.release();
                this.OooO0oo = null;
            }
            ImageReader imageReader = this.OooO0oO;
            if (imageReader != null) {
                imageReader.close();
                this.OooO0oO = null;
            }
        }
    }

    public final void updateScreenSize() {
        WindowManager windowManager;
        if (this.OooOOOo) {
            int rotation = 0;
            this.OooOOo0 = false;
            int[] iArrOooO00o = C4577yf.OooO00o(this.OooO00o);
            int i = iArrOooO00o[0];
            int i2 = iArrOooO00o[1];
            Context context = this.OooO00o;
            if (context != null && (windowManager = (WindowManager) context.getSystemService("window")) != null) {
                rotation = windowManager.getDefaultDisplay().getRotation();
            }
            this.OooOOO = rotation;
            if (this.OooOO0O == i && this.OooOO0o == i2) {
                return;
            }
            this.OooOO0O = i;
            this.OooOO0o = i2;
        }
    }
}

File: app/src/main/java/com/example/myapplication/screencap/ScreenShoterV3.kt
package com.example.myapplication.screencap

import android.annotation.TargetApi
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.Paint
import android.hardware.display.DisplayManager
import android.media.Image
import android.media.ImageReader
import android.media.projection.MediaProjection
import android.media.projection.MediaProjectionManager
import android.os.Handler
import android.os.HandlerThread
import android.util.Log
import android.view.WindowManager
import java.nio.ByteBuffer

/**
 * MediaProjection 截图器 - 参考 ScreenShoterV3.java
 * 通过 VirtualDisplay + ImageReader 实现高效截图
 */
@TargetApi(21)
class ScreenShoterV3 private constructor() {

    companion object {
        private const val TAG = "ScreenShoterV3"

        @Volatile
        private var instance: ScreenShoterV3? = null

        fun getInstance(): ScreenShoterV3 {
            return instance ?: synchronized(this) {
                instance ?: ScreenShoterV3().also { instance = it }
            }
        }
    }

    private var context: Context? = null
    private var resultIntent: Intent? = null
    private var projection: MediaProjection? = null
    private var imageReader: ImageReader? = null
    private var virtualDisplay: android.hardware.display.VirtualDisplay? = null
    private var handlerThread: HandlerThread? = null
    private var handler: Handler? = null

    private var screenWidth = 0
    private var screenHeight = 0
    private var screenDensity = 0
    private var rotation = 0

    private var isInited = false
    private var isCapturing = false

    @Volatile
    private var imageReady = false
    private var lastBitmap: Bitmap? = null
    private var lastImageData: ScreenShotImage? = null

    private val lock = Any()

    /** 初始化 (由 ForScreenShotActivity 回调) */
    fun init(context: Context, resultIntent: Intent) {
        if (isInited) return
        isInited = true
        this.context = context.applicationContext
        this.resultIntent = resultIntent
        this.screenDensity = android.content.res.Resources.getSystem().displayMetrics.densityDpi
    }

    fun init(context: Context, resultIntent: Intent, projection: MediaProjection) {
        if (isInited) return
        isInited = true
        this.context = context.applicationContext
        this.resultIntent = resultIntent
        this.screenDensity = android.content.res.Resources.getSystem().displayMetrics.densityDpi
        this.projection = projection
    }

    fun isInited(): Boolean = isInited

    /** 更新屏幕尺寸 */
    fun updateScreenSize() {
        if (!isInited) return
        imageReady = false
        val ctx = context ?: return
        val wm = ctx.getSystemService(Context.WINDOW_SERVICE) as? WindowManager ?: return
        val display = wm.defaultDisplay
        val metrics = android.util.DisplayMetrics()
        display.getRealMetrics(metrics)
        rotation = display.rotation
        val w = metrics.widthPixels
        val h = metrics.heightPixels
        if (screenWidth == w && screenHeight == h) return
        screenWidth = w
        screenHeight = h
    }

    /** 启动截图 */
    private fun startCapture() {
        releaseCapture() // 只释放 display/reader，保留 projection
        updateScreenSize()

        val ctx = context ?: return
        // 如果 projection 已经由 ScreenCaptureService 传入，直接使用；否则尝试自行获取（低版本兼容）
        if (projection == null && resultIntent != null) {
            val mgr = ctx.getSystemService(Context.MEDIA_PROJECTION_SERVICE) as? MediaProjectionManager
            projection = mgr?.getMediaProjection(-1, resultIntent!!)
        }
        if (projection == null) {
            Log.e(TAG, "No MediaProjection available")
            return
        }

        handlerThread = HandlerThread("imageReaderHandler").also { it.start() }
        handler = Handler(handlerThread!!.looper)

        imageReader = ImageReader.newInstance(screenWidth, screenHeight, android.graphics.PixelFormat.RGBA_8888, 3)
        virtualDisplay = projection?.createVirtualDisplay(
            "screen-mirror",
            screenWidth, screenHeight, screenDensity,
            DisplayManager.VIRTUAL_DISPLAY_FLAG_AUTO_MIRROR,
            imageReader!!.surface, null, null
        )

        isCapturing = true
        // 持续监听，每次有新帧就更新 lastBitmap
        imageReader?.setOnImageAvailableListener({ reader ->
            val image = reader.acquireLatestImage() ?: return@setOnImageAvailableListener
            try {
                val bmp = imageToBitmap(image)
                if (bmp != null) {
                    // 回收旧的 lastBitmap（如果不同对象）
                    val old = lastBitmap
                    lastBitmap = bmp
                    if (old != null && old !== bmp && !old.isRecycled) {
                        old.recycle()
                    }
                    imageReady = true
                }
            } catch (e: Exception) {
                // imageToBitmap 内部已 close image
            }
        }, handler)
    }

    /** 获取截图 (返回 Bitmap) — 每次返回最新帧的拷贝 */
    fun obtainScreenShotBitmap(cropTop: Int = 0, cropBottom: Int = 0): Bitmap? {
        if (!imageReady || virtualDisplay == null) {
            synchronized(lock) {
                startCapture()
                // 等待第一帧
                var tries = 0
                while (!imageReady && tries++ < 100) {
                    try { Thread.sleep(20) } catch (_: InterruptedException) {}
                }
            }
        } else {
            // VirtualDisplay 已在运行，等一小段时间让最新帧刷新
            try { Thread.sleep(50) } catch (_: InterruptedException) {}
        }

        val bmp = lastBitmap ?: return null
        return try {
            if (bmp.isRecycled) null
            else bmp.copy(bmp.config ?: Bitmap.Config.ARGB_8888, false)
        } catch (e: Exception) {
            Log.e(TAG, "Copy bitmap failed", e)
            null
        }
    }

    /** 获取截图 (返回原始数据) */
    fun obtainScreenShotImage(): ScreenShotImage? {
        if (!imageReady || virtualDisplay == null) {
            synchronized(lock) {
                startCapture()
                var tries = 0
                while (!imageReady && tries++ < 100) {
                    try { Thread.sleep(20) } catch (_: InterruptedException) {}
                }
            }
        } else {
            try { Thread.sleep(50) } catch (_: InterruptedException) {}
        }
        return lastImageData
    }

    /** Image -> Bitmap */
    private fun imageToBitmap(image: Image, cropTop: Int = 0, cropBottom: Int = 0): Bitmap? {
        return try {
            val width = image.width
            val height = image.height
            val planes = image.planes
            val buffer = planes[0].buffer
            val pixelStride = planes[0].pixelStride
            val rowStride = planes[0].rowStride
            val rowPadding = (rowStride - pixelStride * width) / pixelStride

            val bmp = Bitmap.createBitmap(width + rowPadding, height, Bitmap.Config.ARGB_8888)
            bmp.copyPixelsFromBuffer(buffer)

            // 不旋转: 游戏是横屏的，保持原始方向
            // 如果有 rowPadding 导致宽度多余，裁剪掉
            val result = if (rowPadding > 0 && width > 0) {
                val cropped = Bitmap.createBitmap(bmp, 0, 0, width, height)
                bmp.recycle()
                cropped
            } else bmp

            // 裁剪
            if (cropTop != 0 || cropBottom != 0) {
                val w = result.width
                val h = result.height
                val cropped = Bitmap.createBitmap(result, 0, cropTop, w, h - cropTop * 2)
                result.recycle()
                cropped
            } else result
        } catch (e: Exception) {
            null
        } finally {
            image.close()
        }
    }

    /** Image -> ScreenShotImage (原始数据) */
    private fun extractImageData(image: Image): ScreenShotImage? {
        return try {
            val planes = image.planes
            val buffer = planes[0].buffer
            val data = ByteArray(buffer.capacity())
            buffer.get(data)

            ScreenShotImage(
                width = image.width,
                height = image.height,
                rotation = rotation,
                rowStride = planes[0].rowStride,
                pixelStride = planes[0].pixelStride,
                data = data
            )
        } catch (e: Exception) {
            null
        } finally {
            image.close()
        }
    }

    /** 释放截图资源（保留 projection） */
    private fun releaseCapture() {
        imageReady = false
        isCapturing = false
        screenWidth = 0
        screenHeight = 0
        rotation = 0

        synchronized(lock) {
            imageReader?.setOnImageAvailableListener(null, null)
            handlerThread?.quit()
            handlerThread = null
            handler?.removeCallbacksAndMessages(null)
            handler = null
            virtualDisplay?.release()
            virtualDisplay = null
            imageReader?.close()
            imageReader = null
            lastBitmap?.let { if (!it.isRecycled) it.recycle() }
            lastBitmap = null
            lastImageData = null
        }
    }

    /** 释放资源 */
    fun recycle() {
        imageReady = false
        isCapturing = false
        isInited = false
        screenWidth = 0
        screenHeight = 0
        rotation = 0

        synchronized(lock) {
            handlerThread?.quit()
            handlerThread = null
            handler?.removeCallbacksAndMessages(null)
            handler = null
            projection?.stop()
            projection = null
            virtualDisplay?.release()
            virtualDisplay = null
            imageReader?.close()
            imageReader = null
        }
    }
}
